plugin-server/src/main/ingestion-queues/session-recording/utils.ts

```diff
@@ -102,10 +105,169 @@ export const getPartitionsForTopic = (
     if (lag < threshold) {
         return 1
     }
 
     return Math.max(0.1, 1 - (lag - threshold) / (threshold * 10))
 }
+
+export async function readTokenFromHeaders(
+    headers: MessageHeader[] | undefined,
+    getTeamFn: (s: string) => Promise<TeamIDWithConfig | null>
+) {
+    const tokenHeader = headers?.find((header: MessageHeader) => {
+        // each header in the array is an object of key to value
+        // because it's possible to have multiple headers with the same key
+        // but, we don't support that. the first truthy match we find is the one we use
+        return header.token
+    })?.token
+
+    const token = typeof tokenHeader === 'string' ? tokenHeader : tokenHeader?.toString()
+
+    let teamIdWithConfig: TeamIDWithConfig | null = null
+
+    if (token) {
+        teamIdWithConfig = await getTeamFn(token)
+    }
+    return { token, teamIdWithConfig }
+}
+
+export const parseKafkaMessage = async (
+    message: Message,
+    getTeamFn: (s: string) => Promise<TeamIDWithConfig | null>
+): Promise<IncomingRecordingMessage | void> => {
+    const statusWarn = (reason: string, extra?: Record<string, any>) => {
+        status.warn('⚠️', 'invalid_message', {
+            reason,
+            partition: message.partition,
+            offset: message.offset,
+            ...(extra || {}),
+        })
+    }
+
+    if (!message.value || !message.timestamp) {
+        // Typing says this can happen but in practice it shouldn't
+        return statusWarn('message value or timestamp is empty')
+    }
+
+    const headerResult = await readTokenFromHeaders(message.headers, getTeamFn)
+    const token: string | undefined = headerResult.token
+    let teamIdWithConfig: null | TeamIDWithConfig = headerResult.teamIdWithConfig
+
+    // NB `==` so we're comparing undefined and null
+    // if token was in the headers but, we could not load team config
+    // then, we can return early
+    if (!!token && (teamIdWithConfig == null || teamIdWithConfig.teamId == null)) {
+        eventDroppedCounter
+            .labels({
+                event_type: 'session_recordings_blob_ingestion',
+                drop_cause: 'header_token_team_missing_or_disabled',
+            })
+            .inc()
+
+        return statusWarn('header_token_team_missing_or_disabled', {
+            token: token,
+        })
+    }
+
+    let messagePayload: RawEventMessage
+    let event: PipelineEvent
+
+    try {
+        messagePayload = JSON.parse(message.value.toString())
+        event = JSON.parse(messagePayload.data)
+    } catch (error) {
+        return statusWarn('invalid_json', { error })
+    }
+
+    const { $snapshot_items, $session_id, $window_id } = event.properties || {}
+
+    // NOTE: This is simple validation - ideally we should do proper schema based validation
+    if (event.event !== '$snapshot_items' || !$snapshot_items || !$session_id) {
+        status.warn('🙈', 'Received non-snapshot message, ignoring')
```

Comment:
let's increment `eventDroppedCounter` too here for visibility. Ideally, I'd love for all drop scenarios to increment this metric.